### ISSUE TITLE
Code re-org: data loading

### ISSUE THREAD

**Issue by grantmerz** [lincc-frameworks/deepdisc#33]
The data loading should be generalized more to deal with different datasets.  Changes can go under /src/deepdisc/data_format/. The way data is loaded in the detectron2 framework is described here [https://detectron2.readthedocs.io/en/latest/tutorials/data_loading.html]

1. A function is built to read in all the metadata from memory and is passed to the DatasetCatalog.register() method to register it.  This function should return a `list[dict]` Right now this function is `get_data_from_json`. It assumes the metadata is saved as a json list of dicts, formatted in the COCO format (see https://detectron2.readthedocs.io/en/latest/tutorials/datasets.html).  Images are not stored in the dictionaries, only filepaths to the images.
2. The train_mapper_cls class will take as input a single item from the `list[dict]`.  It will use the image filepath to read in the image using some function, apply augmentations, store it in the dictionary and slightly change the format of the dictionary metadata.  This new dictionary that contains the image and reformatted metadata is what is fed into the model.  Right now the function to read in images in the train_mapper_cls is the `read_image_hsc` function inside `astrodet/astrodet`.  

Both functions in step 1 and 2 should be easily customizable to deal with different dataset formats.  I think having some general base class for step 1 for reading in data and returning a list[dict] would be useful.  Also a base class for reading in image data would also be useful.  This would address much of issue #30.

**Comment by grantmerz** [comment]
- [ ] Generalize get_data_from_json to class that has a method to read in all metadata and output it in coco dict format
- [ ] Generalize the read_image methods to a class that will take a user specified function, apply some scaling, and return an image based on a key found in the dictionary output from step 1.

**Comment by dougbrn** [comment]
@grantmerz, what are some of the other dataset formats we want to support beyond the json/COCO format?

**Comment by grantmerz** [comment]
For the most part, code should stick with list[dict] COCO format. For example, if you look at `demo_decam.ipynb`, there is a `get_astro_dicts` function that outputs a list[dict] after manually reading in files and formatting everything to COCO format.  Below it, there is a `get_data_from_json` function that reads in a COCO pre-formatted json file and returns a list[dict]. Since these output the same thing, my vision was to have a common class for them. The user would still have to write the code in `get_astro_dicts` since image files may be formatted differently.  But it would be useful to have the option to customize this portion and output a list[dict] that doesn't adhere to COCO format.

**Comment by grantmerz** [comment]
The utility is that if we want to generalize this for other models, we can still use the data loading tools that assume list[dict] but the dict can contain whatever data those other models need

**Comment by dougbrn** [comment]
As a small status update. I've been working on this here: https://github.com/lincc-frameworks/deepdisc/blob/data_loader/src/deepdisc/data_format/file_io.py#L12 

The main idea being that the `DataLoader` class has a `generate_filedict` function which reads in filenames from a given directory. It's generalizable to the extent of being able to handle DECAM and HSC data within the same API. The output is standardized and then can be fed into `generate_dataset_dict` to pass the image and mask files along to user-defined functions. The result is a list of whatever output is produced by the UDF at the moment. I also plan on potentially adding the canned get_astro_dict() functions from the DECAM and HSC notebooks as available on-hand functions here, modified to work with the format being passed along to the UDFs.

The list[dict] format can also be gotten directly from a file, like `get_data_to_json` if it's available.

Plan to keep working on this tomorrow, and actually trying this code will almost certainly not work as I've yet to start using it in any capacity. Let me know if this looks totally off-base from what is being requested.

**Comment by grantmerz** [comment]
I like the way this is structured.  It would be nice to change the self.dataset_dicts to a self.dataset property.  This is to indicate that the class should be able to output data that might not be in dictionary format, for instance a loaded numpy array of all images+metadata.  This output is reformatted in the DataMapper class which feeds directly into the model, so it's ok if the output data from DataLoader isn't necessarily a dictionary (it just needs to be indexable for pytorch backend). Then, have a general method that users can provide a custom function to for loading the dataset.  This way, the user could insert a loading function (like np.load/get_data_from_json).  Maybe the generate_dataset_dict function could be part of a ScarletDataLoader subclass
